### PR TITLE
Refactor generics across Engine

### PR DESCRIPTION
**Pull request by serhiikrivtsov**
Isn't it simpler, or you have any reasons to do it in the way you pick?

### CODE REVIEW

**Comment by caiozullo** [comment]
Hi @KrSerhiy, thanks. 

That looks simpler, however, I'm afraid this makes changes to the code more fragile since changes to the type constraints could be breaking changes to the clients of the framework.

For example, if we remove one of the associated types from the Router protocol in the future, we would have to fix all the call sites (clients). By keeping it as is, we only have to change the internal type constraints, leaving the clients untouched. 

In short, I don't think we can assume that the Router associated types will always match the Flow and other types.

I'm still thinking about it, so let me know if you agree or have other concerns.

**Comment by serhiikrivtsov** [comment]
Hi @caiozullo, you are doing good materials, thanks for that.

Removing one of the associated types from `Router` can't be done without affecting clients, cause you need to replace removed associated type with some concrete type, which may not match to the client's implementation of the protocol. 

Addition of a new placeholder type to the `Flow`, `Game` classes, and to the `startGame(questions: router: correctAnswers: )` function, instead of using the associated type of the `Router`,  in some cases, will be resolved by type inference, and in others - client will need just to specify additional type parameter, in a variable definition. Not too much.

Maybe, the thought, that use of separate placeholder types decouples, in some way, the `Router` from the `Flow`,  has some sense. But now, I can't figure out the real example, where it really comes in handy.

**Comment by caiozullo** [comment]
@KrSerhiy you're right, changes to the `Router` protocol would affect the clients, but only in the types that implement the protocol. This still can be safer and affect fewer clients when compared to changes that might affect method invocations and return types. Type inference should cover most of it though!

That's when types can get in the way – in dynamic languages that would not be a problem.

The `Flow` is internal to the project, so I believe it's fine to use the `Router` associated types there. 

The `Game` is `public` and makes me wonder if it's the correct thing to do. Your changes already affected the outside boundary, so maybe we're looking at the wrong problem here. Maybe the `Game` should act as a `type eraser` and not expose any types. That would give us more flexibility for changes and be safer for clients. What do you think?

**Comment by serhiikrivtsov** [comment]
@caiozullo, I guess, in your implementation, you are using the technic similar to the type eraser, but it's not exactly the type eraser. The purpose of the type eraser to move the emphasis from the concrete type that implements the protocol to the concrete associated types:

```swift
// the router variable erases concrete type that implements Router protocol and cares only about concrete parameter types
var router: AnyRouter<Question<String>, Set<String>> = AnyRouter(NavigationControllerRouter(...))
router = AnyRouter(AnotherRouter(...)) // AnotherRouter is an implementation of the Router protocol in which also Question == Question<String> and Answer == Set<String>.
```

And in your implementation, a concrete type, that implements the `Router` protocol, isn't erased nor from the `Game`, either from the `Flow`.

But your comment inspired me to try the implementation with exactly the type eraser of the `Router` protocol, and I think that it's what we are looking for. I have already committed those changes. What do you think?

I have one more refactoring suggestion. We can replace the `startGame(questions: router: correctAnswers:)` function whit initializer and a `start` method of the `Game` class, and the `start` method will use a factory method to create a new `Flow` instance inside. And also we can move the `scoring(_: correctAnswers:)` function inside the `Game` class if you don't consider that scoring shouldn't belong to the `Game` class responsibilities (I don't). That will remove the `Question`, `Answer` placeholder types from those functions, and makes them more readable. If you don't exactly get what I mean, I can provide the implementation. Let me also know what do you think about this.

**Comment by serhiikrivtsov** [comment]
To @caiozullo. I have already implemented the refactoring suggestion from the previous comment to be more clear.

File: Engine/QuizEngineTests/GameTest.swift [hunk 1]
```diff
@@ -9,12 +9,12 @@ import QuizEngine
 class GameTest: XCTestCase {
     
     let router = RouterSpy()
-    var game: Game<String, String, RouterSpy>!
+    lazy var game = Game(questions: ["Q1", "Q2"], router: router, correctAnswers: ["Q1": "A1", "Q2": "A2"])
```
Comment (caiozullo) [inline Engine/QuizEngineTests/GameTest.swift]: Like this! Was it possible before Swift 4?

**Comment by caiozullo** [comment]
Hey @KrSerhiy sorry for the late response – been super busy here.

I think my analogy with the `type eraser` was a bad one and generated some miscommunication, sorry. 

What I was trying to say is to find a way to remove the generic types from the Game. Something like:

```
public class Game {
    private let flow: Any
    
    init(flow: Any) {
        self.flow = flow
    }
}
```

By making the property private, I don't mind what type it is for now. I just care about saving the clients to deal with the generics. It's a pain!

The AnyRouter you implemented looks neat but I don't think we need it. I'm happy to have your previous commit, with the Flow generic types simplified with types taken from the Router protocol. That's an internal implementation and should be fine.

About the start game inside the Game class, I think it's a good idea. Any reason why you prefer an initializer and a start method for the Game though? I think a `public static start` method could do fine and avoid clients having to have the extra step to create, then start. I'm fine with both actually but want to know your opinion.

**Comment by serhiikrivtsov** [comment]
To @caiozullo. I prefer the `start` and `scoring` methods instead of the free functions, cause, in this situation, it allows us to remove the type parameters from the definitions of the functions and use the type parameters of the `Game` class for the methods:

```swift
// INSTEAD OF:
// with type parameters for class
public class Game <Question, Answer, R: Router> where R.Question == Question, R.Answer == Answer {
...
}

// with function's type parameters
public func startGame<Question, Answer: Equatable, R: Router>(questions: [Question], router: R, correctAnswers: [Question: Answer]) -> Game<Question, Answer, R> {...}

// with function's type parameters
private func scoring<Question, Answer: Equatable>(_ answers: [Question: Answer], correctAnswers: [Question: Answer]) -> Int {...}

// WE HAVE THIS:
// only type parameters for class
public class Game <Question, Answer: Equatable, R: Router> where R.Question == Question, R.Answer == Answer {
...
// without function's type parameters
public init(questions: [Question], router: R, correctAnswers: [Question: Answer]) {...} 

public start() {...}

// without function's type parameters
private func scoring(_ answers: [Question: Answer], correctAnswers: [Question: Answer]) -> Int {...}
}
```

File: Engine/QuizEngineTests/GameTest.swift (hunk 1 above)
Comment (serhiikrivtsov) [reply Engine/QuizEngineTests/GameTest.swift]: If you mean `lazy` properties, then it was.

Comment (caiozullo) [reply Engine/QuizEngineTests/GameTest.swift]: No, I meant using the initializer in the lazy var

Comment (serhiikrivtsov) [reply Engine/QuizEngineTests/GameTest.swift]: I guess, also yes.

**Comment by serhiikrivtsov** [comment]
So, @caiozullo,  in your opinion, the better solution will be to leave `Game` class, as it was in the beginning, and to change the `Flow` class in the way to use the `Router`'s associated types instead of the separate type parameters?

**Comment by serhiikrivtsov** [comment]
To @caiozullo. I understand, that this is an example project for your video course, and there is no real reason to do this refactoring stuff, only if you are just interested in it. So, if you are not interested in it, or you haven't really time for that, or something else, you can just say it, and I will not take your time.

**Comment by caiozullo** [comment]
Hi @serhiikrivtsov, yes I'd prefer to leave the Game class as it was. We have plans to change this class for future videos and it will solve these issues.

If you revert it back, I'm happy to merge the initial changes you did for the Flow.

Thanks!

**Comment by serhiikrivtsov** [comment]
To @caiozullo. I have done it. Please check it, to be sure, that I understood you correctly.

**Comment by caiozullo** [comment]
@serhiikrivtsov awesome! Merging.

**Comment by caiozullo** [comment]
@serhiikrivtsov we're really glad for people to use this codebase to share knowledge. We're openly inviting people to participate but we're still going to be reviewing and maybe not all changes will be accepted/merged. 

I might not have time to check this every day but will be looking every time possible.

You've done great work, thank you very much for sharing!
